packages/compass-collection/src/modules/collection-tab.ts

```diff
@@ -698,32 +696,114 @@ export const cancelSchemaAnalysis = (): CollectionThunkAction<void> => {
+ * Moves fieldPath from object property to object key.
+ */
+function transformFakerSchemaToObject(
+  fakerSchema: FakerSchemaMapping[]
+): FakerSchema {
+  const result: FakerSchema = {};
+
+  for (const field of fakerSchema) {
+    const { fieldPath, ...fieldMapping } = field;
+    result[fieldPath] = {
+      ...fieldMapping,
+      mongoType: fieldMapping.mongoType,
+    };
+  }
+
+  return result;
+}
+
+/**
+ * Checks if the method exists and is callable on the faker object.
+ *
+ * Note: Only supports the format `module.method` (e.g., `internet.email`).
+ * Nested modules or other formats are not supported.
+ * @see {@link https://fakerjs.dev/api/}
+ */
+function isValidFakerMethod(fakerMethod: string): boolean {
+  const parts = fakerMethod.split('.');
+
+  // Validate format: exactly module.method
+  if (parts.length !== 2) {
+    return false;
+  }
+
+  const [moduleName, methodName] = parts;
+
+  try {
+    const fakerModule = (faker as unknown as Record<string, unknown>)[
+      moduleName
+    ];
+    return (
+      fakerModule !== null &&
+      fakerModule !== undefined &&
+      typeof fakerModule === 'object' &&
+      typeof (fakerModule as Record<string, unknown>)[methodName] === 'function'
+    );
+  } catch {
+    return false;
+  }
+}
+
+/**
+ * Validates a given faker schema against an input schema.
+ *
+ * - Transforms LLM array format to keyed object structure
+ * - Validates the `fakerMethod` for each field, marking it as unrecognized if invalid
+ * - Adds any unmapped input schema fields to the result with an unrecognized faker method
+ *
+ * @param inputSchema - The schema definition for the input, mapping field names to their metadata.
+ * @param fakerSchemaArray - The array of faker schema mappings from LLM to validate and map.
+ * @param logger - Logger instance used to log warnings for invalid faker methods.
+ * @returns A keyed object of validated faker schema mappings, with one-to-one fields with input schema.
+ */
 const validateFakerSchema = (
-  fakerSchema: MockDataSchemaResponse,
+  inputSchema: Record<string, FieldInfo>,
+  fakerSchemaArray: FakerSchemaMapping[],
   logger: Logger
-) => {
-  return fakerSchema.fields.map((field) => {
-    const { fakerMethod } = field;
-
-    const [moduleName, methodName, ...rest] = fakerMethod.split('.');
-    if (
-      rest.length > 0 ||
-      typeof (faker as any)[moduleName]?.[methodName] !== 'function'
-    ) {
-      logger.log.warn(
-        mongoLogId(1_001_000_372),
-        'Collection',
-        'Invalid faker method',
-        { fakerMethod }
-      );
-      return {
-        ...field,
+): FakerSchema => {
+  // Transform to keyed object structure
+  const fakerSchemaRaw = transformFakerSchemaToObject(fakerSchemaArray);
+
+  const result: FakerSchema = {};
+
+  // Process all input schema fields in a single O(n) pass
+  for (const fieldPath of Object.keys(inputSchema)) {
+    const fakerMapping = fakerSchemaRaw[fieldPath];
```

Comment:
Now that we're removing the probability from the backend response model, we should just get the probability from the input schema

```suggestion
    const fakerMapping = { ...fakerSchemaRaw[fieldPath], probability: inputSchema[fieldPath] }
```

Or something similar.